addons/controllers/clusterbootstrap_controller.go

```diff
@@ -1464,3 +1539,157 @@ func (r *ClusterBootstrapReconciler) reconcileClusterProxyAndNetworkSettings(clu
 
 	return nil
 }
+
+func (r *ClusterBootstrapReconciler) clusterIsReadyForDeletion(cluster *clusterapiv1beta1.Cluster, log logr.Logger) (bool, error) {
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	err := r.Client.Get(r.context, client.ObjectKeyFromObject(cluster), clusterBootstrap)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "failed to lookup clusterbootstrap")
+		return false, err
+	}
+	if apierrors.IsNotFound(err) {
+		log.Info("cluster bootstrap not found")
+	} else {
+		remoteClient, err := util.GetClusterClient(r.context, r.Client, r.Scheme, clusterapiutil.ObjectKey(cluster))
+		if err != nil {
+			return false, err
+		}
+		if hasPackageInstalls(r.context, remoteClient, cluster, r.Config.SystemNamespace, clusterBootstrap.Spec.AdditionalPackages, log) {
```

Comment:
I'm not sure whether we need this `hasPackageInstalls` check before running `removeAdditionalPackageInstalls`. As we will just continue is all the errors are `apierrors.IsNotFound`.

Reply:
hasPackageInstall is wha actually returns 'false' for clusterIsReadyForDeletion (makeClusterReadyForDeletion). 
So we do need to keep the check, and I think we probably want to place the call to removeAdditionaPackageInstalls inside the if statement so we can return a false, nil right away. 
i.e. 
if we find packages we ask to remove them and signal that cluster is not ready for deletion.

Reply:
done